core/src/commons/Data.cpp

```diff
@@ -174,22 +175,41 @@ void Data::get_all_values(std::vector<double>& all_values,
    // fill with [0, 1,..., samples.size() - 1]
   std::iota(index.begin(), index.end(), 0);
   // sort index based on the split values (argsort)
-  std::sort(index.begin(), index.end(), [&](const size_t& lhs, const size_t& rhs)
-    {return all_values[lhs] < all_values[rhs];});
+  if (contains_nan()) {
+    // the NaN comparison places all NaNs at the beginning
+    // this will be slightly slower if there are no NaNs, thus the branch
+    std::stable_sort(index.begin(), index.end(), [&](const size_t& lhs, const size_t& rhs) {
```

Comment:
Did you test the option of alway using `std::stable_sort` and notice a difference in runtime? It looks slower from benchmarks, but not too much slower, and I wonder if we could simplify here and always use `std::stable_sort`. If feels more robust from a cross-platform perspective, as it is a bit of an accident that for non-NaNs, ties are handled the same way in different sort implementations.

Reply:
Yes, there is a difference, but not humongous. Will simplify it, as it looks strange from a robustness perspective.

```R
# A large mix continuous/discrete forest (num.threads=12)
set.seed(1)
n <- 1000000
p <- 20
X <- matrix(rnorm(n * p), n, p)
X[1:n, 11:p] <- sample(1:200, n*10, replace = T)
W <- rbinom(n, 1, 0.5)
Y <- pmax(X[, 1], 0) * W + X[, 2] + pmin(X[, 3], 0) + rnorm(n)
system.time(c.forest <- causal_forest(X, Y, W, seed = 1, num.trees = 500))

# master with only small q and std::sort
     user    system   elapsed 
14555.096    62.938  1398.976 

# master with only small q and std::stable_sort
15518.996    62.335  1472.956
```

Reply:
Just adding a post merge note that the main (minor) slowdown in causal_benchmark.R arises from the change to always use stable sorting:

```
MIA @ ce2cbeb (just before always use stable sort)

         oob               test              stdratio
Y.error   "0.048 +/- 0.001" "0.049 +/- 0.001" "2.667 +/- 0.110"
tau.error "0.010 +/- 0.001" "0.009 +/- 0.001" "1.298 +/- 0.093"
              user.self          sys.self          elapsed
Y.time*        "16.586 +/- 0.067" "0.070 +/- 0.011" "16.677 +/- 0.078"
Y.time.pred   "2.486 +/- 0.017"  "0.051 +/- 0.007" "2.545 +/- 0.018"
Y.time.ci     "5.432 +/- 0.049"  "0.050 +/- 0.005" "5.494 +/- 0.055"
tau.time*      "15.723 +/- 0.055" "0.042 +/- 0.004" "15.790 +/- 0.056"
tau.time.pred "2.531 +/- 0.039"  "0.029 +/- 0.004" "2.565 +/- 0.040"
tau.time.ci   "4.841 +/- 0.073"  "0.036 +/- 0.004" "4.890 +/- 0.077"
```

```
MIA @ f948d4b (always use stable sort)

          oob               test              stdratio
Y.error   "0.048 +/- 0.001" "0.049 +/- 0.001" "2.667 +/- 0.110"
tau.error "0.010 +/- 0.001" "0.009 +/- 0.001" "1.298 +/- 0.093"
              user.self          sys.self          elapsed
Y.time*        "17.205 +/- 0.048" "0.075 +/- 0.010" "17.302 +/- 0.058"
Y.time.pred   "2.492 +/- 0.023"  "0.055 +/- 0.006" "2.550 +/- 0.025"
Y.time.ci     "5.485 +/- 0.038"  "0.054 +/- 0.004" "5.546 +/- 0.039"
tau.time*      "16.035 +/- 0.071" "0.048 +/- 0.004" "16.106 +/- 0.075"
tau.time.pred "2.473 +/- 0.023"  "0.027 +/- 0.003" "2.502 +/- 0.024"
tau.time.ci   "4.775 +/- 0.076"  "0.028 +/- 0.002" "4.808 +/- 0.078"
```

```
MIA @ latest

          oob               test              stdratio
Y.error   "0.048 +/- 0.001" "0.049 +/- 0.001" "2.667 +/- 0.110"
tau.error "0.010 +/- 0.001" "0.009 +/- 0.001" "1.298 +/- 0.093"
              user.self          sys.self          elapsed
Y.time*        "17.259 +/- 0.078" "0.073 +/- 0.010" "17.369 +/- 0.101"
Y.time.pred   "2.519 +/- 0.024"  "0.055 +/- 0.007" "2.583 +/- 0.024"
Y.time.ci     "5.510 +/- 0.064"  "0.055 +/- 0.005" "5.574 +/- 0.066"
tau.time*      "16.049 +/- 0.079" "0.051 +/- 0.008" "16.139 +/- 0.095"
tau.time.pred "2.555 +/- 0.051"  "0.025 +/- 0.003" "2.582 +/- 0.053"
tau.time.ci   "4.801 +/- 0.048"  "0.029 +/- 0.002" "4.834 +/- 0.050"
```

The [causal_benchmark.txt](https://gist.github.com/erikcs/77ff794c1ea323605bb091dffe85ee74) linked to in in this issue summary contained a wrong entry, sorry.